src/compiler/binder.ts

```diff
@@ -278,6 +278,15 @@ namespace ts {
                     Debug.assert(isWellKnownSymbolSyntactically(nameExpression));
                     return getPropertyNameForKnownSymbolName(idText((<PropertyAccessExpression>nameExpression).name));
                 }
+                if (isPrivateIdentifier(name)) {

```

Comment:
@sandersn responding to your [question](https://github.com/microsoft/TypeScript/pull/30829#pullrequestreview-262682575)

> Also still need some js tests, particularly to make sure that this-property assignments in the constructor give a reasonable error suggesting a private property declaration.

> (I don't think GH provides threaded replies for comments that aren't on a line of code)

Your example points to something I messed up, We don't currently error on this *in JS files*:

```js
class A {
    constructor () {
        this.#foo = 3
    }
}
```

**Q1**:
Could you point me in the right direction for fixing this? I suspect there is a difference in the binder here depending on the file extension:
- If I change the extension to .ts, in checkPropertyAccessExpressionOrQualifiedName the type does *not* contain a `foo` member.
- If the file extension is .js, in checkPropertyAccessExpressionOrQualifiedName  the type *does* contain a `foo` member. I'm just not finding where this branch in logic lives.

**Q2**: do you have other suggestions for things that are different between a JS file and a TS file where everything is `any`? I can't think of other differences.

Thanks for your help!

Reply:
Re Q1: I think we can do a check for private identifiers in `bindThisPropertyAssignment`. I think that's where we add properties to the `this` type in the constructor (but only for JS files)

Reply:
Static property assignment should be disallowed too:

```js
class A { }
A.#no = 3
```

Pre-property declarations, I think this was the only way to make static properties, right?

Reply:
prototype property assignment shouldn't work, although I think there is very little chance it will:

```js
function A() { }
A.prototype = {
  #x: 1
}
class B { }
// this is wrong in several ways!
B.prototype = {
  #y: 2
}
```

Reply:
And exports/prototype property assignment?  

```js
exports.#nope = 1
function A() { }
A.prototype.#no = 2
```

Reply:
consolidating https://github.com/microsoft/TypeScript/pull/30829/files/f5936ed28c50cf89a803a1556f9f28fed43e7d5f#r305455175